src/main/kotlin/dev/restate/sdktesting/tests/KafkaAndWorkflowAPITest.kt

```diff
@@ -0,0 +1,142 @@
+import org.apache.kafka.clients.producer.Producer
+import org.apache.kafka.clients.producer.ProducerRecord
+import org.assertj.core.api.Assertions.assertThat
+import org.awaitility.kotlin.await
+import org.awaitility.kotlin.withAlias
+import org.junit.jupiter.api.Test
+import org.junit.jupiter.api.extension.RegisterExtension
+import org.junit.jupiter.api.parallel.Execution
+import org.junit.jupiter.api.parallel.ExecutionMode
+
+private const val COUNTER_TOPIC = "counter"
+private const val EVENT_HANDLER_TOPIC = "event-handler"
+
+private fun kafkaClusterOptions(): RestateConfigSchema {
+  return RestateConfigSchema()
+      .withIngress(
+          IngressOptions()
+              .withKafkaClusters(
+                  listOf(
+                      KafkaClusterOptions()
+                          .withName("my-cluster")
+                          .withBrokers(listOf("PLAINTEXT://kafka:9092")))))
+}
+
+class KafkaAndWorkflowAPITest {
+
+  @Workflow
+  class MyWorkflow {
+    @Workflow suspend fun run(ctx: WorkflowContext, myTask: String) = "Run $myTask"
+  }
+
+  companion object {
+    @RegisterExtension
+    val deployerExt: RestateDeployerExtension = RestateDeployerExtension {
+      withContainer("kafka", KafkaContainer(COUNTER_TOPIC, EVENT_HANDLER_TOPIC))
+      withConfig(kafkaClusterOptions())
+      withEndpoint(Endpoint.bind(MyWorkflow()))
+    }
+  }
+
+  @Test
+  @Execution(ExecutionMode.CONCURRENT)
+  fun handleEventInEventHandler(
+      @InjectAdminURI adminURI: URI,
+      @InjectContainerPort(hostName = "kafka", port = KafkaContainer.EXTERNAL_PORT) kafkaPort: Int,
+      @InjectClient ingressClient: Client
+  ) = runTest {
+    val counter = UUID.randomUUID().toString()
+
+    // Create subscription
+    val subscriptionsClient =
+        SubscriptionApi(ApiClient().setHost(adminURI.host).setPort(adminURI.port))
+    subscriptionsClient.createSubscription(
+        CreateSubscriptionRequest()
+            .source("kafka://my-cluster/$EVENT_HANDLER_TOPIC")
+            .sink(
+                "service://${KafkaAndWorkflowAPITestMyWorkflowHandlers.Metadata.SERVICE_NAME}/run")
+            .options(mapOf("auto.offset.reset" to "earliest")))
+
+    val keyMessages = linkedMapOf("a" to "1", "b" to "2", "c" to "3")
+
+    // Produce message to kafka
+    produceMessageToKafka(
+        "PLAINTEXT://localhost:$kafkaPort",
+        EVENT_HANDLER_TOPIC,
+        keyMessages.map { it.key to Json.encodeToString(it.value) })
+
+    // Now assert that those invocations are stored there, let's do this twice just for the sake of.
+    for (keyMessage in keyMessages) {
+      await withAlias
+          "Workflow invocations from Kafka" untilAsserted
+          {
+            assertThat(
+                    KafkaAndWorkflowAPITestMyWorkflowClient.fromClient(
+                            ingressClient, keyMessage.key)
+                        .workflowHandle()
+                        .attachSuspend()
```

Comment:
What does the "Suspend" suffix mean here?

Reply:
it's the kotlin client API, it's to indicate that those operations needs to be executed as suspendable. It's not Restate lingo.

Reply:
Ah ok. I assume that's the idiomatic way to do it in Kotlin?